tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/RecordingZ3SolverContext.scala

```diff
@@ -0,0 +1,207 @@
+package at.forsyte.apalache.tla.bmcmt.smt
+
+import at.forsyte.apalache.tla.bmcmt.profiler.SmtListener
+import at.forsyte.apalache.tla.bmcmt.{Arena, ArenaCell}
+import at.forsyte.apalache.tla.lir.TlaEx
+import com.typesafe.scalalogging.LazyLogging
+
+
+object RecordingZ3SolverContext {
+  def apply(parentLog: Option[SmtLog], config: SolverConfig): RecordingZ3SolverContext = {
+    val context = new RecordingZ3SolverContext(parentLog, config)
+    parentLog match {
+      case Some(log) =>
+        log.replay(context.solver)
+
+      case None => ()
+    }
+    context
+  }
+}
+
+/**
+  * A wrapper around SolverContext that records SMT operations. The log is organized hierarchically,
+  * so a differential diff may be taken.
+  *
+  * @param parentLog the log of the parent context
+  * @param config solver configuration
+  *
+  * @author Igor Konnov
+  */
+@SerialVersionUID(700L)
+class RecordingZ3SolverContext private (parentLog: Option[SmtLog], val config: SolverConfig)
+    extends SolverContext with Serializable with LazyLogging {
+
+  import SmtLog._
+
+  private val solver = new Z3SolverContext(config)
+
+  /**
+    * The sequence of logs (the last added element is in the head),
+    * one per context, except the last log, which is maintained in lastLog.
+    * Every list in logStackRev is stored in the reverse order.
+    */
+  private var logStackRev: List[List[Record]] = List()
```

Comment:
Just a nit on naming, but if, conceptually, this is a stack, then it seems like needlessly exposing an implementation detail to prefix the name with `rev`, cause we'd expect the last added item to be on the top of the stack.

Reply:
It is indeed a confusing name. Changed it to `stackOfInverseLogs` to stress that every log on the stack is in the inverse order.

Reply:
That helps. Thanks!